third_party/proton/dialect/include/Dialect/Proton/IR/ProtonOps.td

```diff
@@ -57,9 +67,60 @@ def TT_RecordOp : TT_Proton_Op<"record", [DeclareOpInterfaceMethods<MemoryEffect
     ins BoolAttr: $isStart,
     ConfinedAttr<I32Attr, [IntNonNegative]>:$regionId,
     DefaultValuedAttr<MetricAttr, "Metric::CYCLE">:$metric,
-    DefaultValuedAttr<GranularityAttr, "Granularity::WARPGROUP">:$granularity
+    DefaultValuedAttr<GranularityAttr, "Granularity::WARPGROUP">:$granularity,
+    DefaultValuedAttr<StrategyAttr, "Strategy::CIRCULAR">:$strategy
   );
+
   let assemblyFormat = " `(` operands `)` attr-dict";
 }
 
+def PT_CircularRecordOp : PT_Op<"circular_record", [DeclareOpInterfaceMethods<MemoryEffectsOpInterface>]> {
+  let summary = "Record a GPU hardware event into a circular buffer";
```

Comment:
NIT: it's not always a "hardware" event, we can record any even into the circular buffer, including those analyzed purely from static analysis